### ISSUE TITLE
Hscan in Redis with 1,000,000 records is taking longer than usual

### ISSUE THREAD

**Issue by chinmaym7430** [xetorthio/jedis#1338]
I am trying to scan a key space having about a million keys by using pattern scan. I am using Jedis 2.8.1 and spring-data-redis 1.7.2 . The amount of time it takes to scan the entire key set thousand times is about 5 minutes. The same thing takes about 10s when done on MySQL. I think I might be doing something wrong since Redis is supposed to be faster than SQL-based databases.

Here are the steps for execution:
1. Set up the spring configuration as mentioned below.
2. This is the method I am using for scanning the keys(Hscan):

```
	@Override
	public List<Object> findByPattern(String pattern) {
		List<Object> listOfObjects = new ArrayList<>();
		Cursor<Entry<Object, Object>> cursorMap = redisTemplate.boundHashOps(SNAPSHOT_MODULE_KEY)
				.scan(ScanOptions.scanOptions().match(pattern).count(100).build());
		if (cursorMap != null) {
			while (cursorMap.hasNext()) {
				listOfObjects.add(cursorMap.next().getKey());
			}
			return listOfObjects;
		} else {
			return null;
		}
	}
```
3. Here is the code by in which I am calling the scan method:

`List<Object> listOfSnapshotModulesKeys = snapshotRepo.findByPattern(snapshotPattern);`

> NOTE: The key for each row is a combination of four separate strings. The pattern that I use for searching(the variable snapshotPattern) is a combination of  three strings. For example, 
>  KEY: 'Str1:Str2:Str3:Str4'
> PATTERN: '*:Srt5:Str6:Str7'

> Also, I am not using connection pooling or pipelining. I have turned off pooling since I was facing an error otherwise. This error has been mentioned in issue: [https://github.com/xetorthio/jedis/issues/918#issuecomment-229922190](url)

**Redis / Jedis Configuration**
![image](https://cloud.githubusercontent.com/assets/18327469/16608122/5b2fa3ae-4366-11e6-8e44-ae28625afae5.png)

**Jedis version**:
2.8.1
 **Redis version**:
3.0.501
 **Java version**:
1.7.0_79
**Spring-data-redis version**:
1.7.2.RELEASE

@marcosnils, @deepakpol, @mp911de: What can I do to make the search faster?

**Comment by mp911de** [comment]
How does performance change if you adjust the batch size? You're using a size of `100` that results in 10.000 network calls.

**Comment by chinmaym7430** [comment]
@mp911de : Thanks for the suggestion. I tried changing the COUNT option.
Here are the results:

- COUNT(100) : Time taken is 5m 15s

- COUNT(10000): Time taken is 4m 17s

- COUNT(50): Time taken is 7m 1s.

Still, the performance is not good. What should I do next?

**Comment by sheinbergon** [comment]
@chinmaym7430  Just to make sure and without delving into the details , you are aware that SCAN is not efficient as  SQL RDBMS index scan by design , right ?

Match patterns are applied against each scanned result output after they've been emitted . This means that first the database emits everything in its current iteration and then applies the match pattern  on top of it 

Do all of iterations of you method take the same time to complete ? Maybe it's a Garbage collection tweaking issue?

Perhaps You can write a simple for loop in bash to test directly against the redis locally ( via redis-cli , I believe the latency of re-running redis-cli and opening the connection should not have a great effect on the results with COUNT size of 10000 and above   ) and see how much time does each iteration takes , regardless of the java integration.I believe it'll be a good starting point to see differences

**Comment by chinmaym7430** [comment]
@sheinbergon After debugging the code, I found out the scan itself is very fast. The actual time is being taken in the part where I iterate through the Cursor and store the results.
```
Cursor<Entry<Object, Object>> cursorMap = redisTemplate.boundHashOps(SNAPSHOT_MODULE_KEY)
				.scan(ScanOptions.scanOptions().match(pattern).count(100).build());
		
		if(cursorMap.hasNext()){       //Iteration
			setOfObjects.add(cursorMap.next().getKey());
			while(cursorMap.getPosition()!=0){
				setOfObjects.add(cursorMap.next().getKey());
			}
			return setOfObjects;
		}else{
			return null;
		}
```
Simply running the scan option didn't take any time. The methods `cursorMap.hasNext()` and `cursorMap.next.getKey()` are time consuming.
Can you suggest a faster method to retrieve the keys from the Cursor that is being returned?

**Comment by sheinbergon** [comment]
@chinmaym7430 This has to do with system design :) can you shed some more light one what is you are trying to achieve in the big picture

**Comment by mp911de** [comment]
The very first call to `Cursor.hasNext` invokes the `SCAN` command. Calling `.scan(…)` just creates a cursor. Spring Data Redis performs on each `.next()` call a conversion from `byte[]` to `String`.

I vote also for reproducing that case with `redis-cli` and then pure Jedis usage to narrow down the source where time is spent.

**Comment by chinmaym7430** [comment]
@sheinbergon Ok, I'll try to explain the big picture in a simple way.
I have a HashMap 'Map1'. Each row in the map has a key-value pair. The key is composed of four strings and the value is an object containing a number of attributes(strings).
The key is actually composed of attributes through which a user can search the database. When a user provides the search criteria, I concat them and this concatenated string is(or should be) one of the keys in the HashMap.
The problem is that the user can provide query criteria such that the entire key can't be formed using them. He can provide lesser number of criterium. Hence, I have to take the input and search keys in the db that have the same pattern as that given by the user.
Once the keys are found, I fetch the corresponding values and return the values to the user.

It is basically a simple search operation using regex patterns. The only constraint is time. I am trying to find a way so that the time to search they keys is minimized.
If the user provides the entire key for searching, there is no problem at all. But I can't do that since the search criteria are to be kept flexible. So I have to search using patterns.
Is there any way I can do this faster?

**Comment by chinmaym7430** [comment]
@mp911de Thanks for the update. So I am back to square one :(

**Comment by sheinbergon** [comment]
@mp911de  yes that sounds right . sorry for missing out on that , my experience with spring-data-redis is very minimal

**Comment by sheinbergon** [comment]
@chinmaym7430  you can do a minimal experiment with redis-cli and loop via bash scripts as suggested earlier. If you lack the knowledge , I can help you with that

**Comment by chinmaym7430** [comment]
@sheinbergon Yeah sure. It would be great if you could help me.

**Comment by sheinbergon** [comment]
@chinmaym7430  will take me a day or two , but I'll get back to you

**Comment by sheinbergon** [comment]
@chinmaym7430  There you go

Bash code to generate a hash with a million entries ( takes just under a minute to complete , thanks to bash streaming capabilites :

```
cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 32  | head -n 1000000  | while read UNIQUE ; do echo " HMSET htest '$UNIQUE' '$UNIQUE' "  ; done  | redis-cli -h localhost -p 6379 --pipe
```

Bash code to scan entire hash for a 3 letter suffix pattern ( you used suffix pattern in your description :
```
COUNT=100 ;ITERATIONS=0 ; CURSOR=5 ; START=`date +%s` ; while [[ $CURSOR -ne 0 ]]  ; do SCAN_RESULT=(`redis-cli hscan htest $CURSOR MATCH *Rdm COUNT $COUNT`) ; CURSOR=${SCAN_RESULT[0]} ; ITERATIONS=$((ITERATIONS+1)) ; done ; END=`date +%s` ; echo $((END-START)) " seconds , $ITERATIONS iterations"
```

COUNT=100 : 19  seconds , 3729 iterations
COUNT=1000 : 3  seconds , 375 iterations
COUNT=10000 : 1  seconds , 38 iterations
COUNT=100000 : 1  seconds , 4 iterations

**Comment by sheinbergon** [comment]
@chinmaym7430 What is the expected match count ? meaning - how many entries  are you expecting to match from a hashset of 1M entries ?

Also , please run the when both the redis and application server reside on the same machine

**Comment by chinmaym7430** [comment]
I am expecting that

**Comment by chinmaym7430** [comment]
On Jul 7, 2016 3:35 PM, "Chinmay Mukhopadhyay" <chinmay7430@gmail.com>
wrote:

>

> I will get back to you in a couple of days.
>
> On Jul 7, 2016 3:05 AM, "Idan Sheinberg" <notifications@github.com> wrote:

>> <https://github.com/chinmaym7430>

>> @chinmaym7430 <https://github.com/chinmaym7430> What is the expected
match count ? meaning - how many entries are you expecting to match from a
hashset of 1M entries ?
>>
>> Also , please run the when both the redis and application server reside
on the same machine
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
<https://github.com/xetorthio/jedis/issues/1338#issuecomment-230914435>, or
mute the thread
<https://github.com/notifications/unsubscribe/ARenrfxAvRuBWnmAbHYAxjXq7byu8Okfks5qTB-mgaJpZM4JFwu4>
.

**Comment by chinmaym7430** [comment]
Please ignore the previous two comments. I will get back to in a couple of days.

**Comment by deepakpol** [comment]
@chinmaym7430 the time you mention is substantial, have you ran some checks through `redis-cli` to ensure your deployment/infrastructure, redis config is not causing the latency? A good source of factors that affect and tools to verify - http://redis.io/topics/latency

**Comment by sheinbergon** [comment]
@deepakpol  I've provide @chinmaym7430  with the means of testing hscan performance locally via redis-cli , waiting for his response,findings

**Comment by chinmaym7430** [comment]
Sorry for the delayed response. As you guys suggested, I tested the performance locally via redis-cli. **The results:**
To scan the 1 million records, redis takes 0.96s when the count is set to 1000000. 
Now as per my requirement, I have to scan the db 273 times and hence the total time amounts to around 4 and a half minutes.
@sheinbergon You are absolutely correct. For a complete scan, it doesn't take more that 1 second. But I have to scan it iteratively for about 300 times. That is why the time taken increases.
Given this, I don't think the performance can be optimized any further. Can you guys suggest some other ideas via which the time taken can be decreased?
Something like if I can process the scan parallelly and not consecutively. Can this be done?

**Comment by mp911de** [comment]
Reducing network roundtrips is in general a good idea. Have you tried to adjust your batch size to a value you end up with 28 instead of 273 times?

**Comment by sheinbergon** [comment]
@chinmaym7430  please refer to https://github.com/xetorthio/jedis/issues/1338#issuecomment-230912584 ( I've put a lot of effort into it ) . doing more than 3500 iterations took me 19 seconds ( for COUNT=100). it makes NO SENSE that for you it takes 4 and a half minutes to cover the same hash.

By 'scan it iteratively '  you mean run 300 different full scan cycles or are you forced to use a small value for COUNT ? if so , why ?

**Comment by chinmaym7430** [comment]
I have to perform  300 full scans. I don't have to use a small count value. In you comment you mentioned that you did one complete scan of the hash and it took you one second. That is exactly what i did. One entire scan takes just one second as you mentioned.  But I have to do the full scan cycle 300 times.  That is what the requirement is.

**Comment by deepakpol** [comment]
@chinmaym7430 so do you repeat the scan 273 times (for some application requirement)? couldn't get the context on why it has to be repeated and how can the number 273 be fixed?

**Comment by sheinbergon** [comment]
@chinmaym7430  im with @deepakpol - I don't understand it either. Anyways -  running  a complete scan cycle ( exhusting a new cursor )  300 times should be compared to running 300 Index scans in an RDBMs  , yet you compared it to a single query back in the original issue. Maybe there's something we are missing here....

**Comment by chinmaym7430** [comment]
@deepakpol The scan has to be repeated based on an application requirement. The number 273 isn't fixed. It;s just an example. The data setup I'm testing with now has 273 patterns. So I have to repeat the scan 273 times. This number can vary based on the application requirement.
Something like this:

```
List<String> patterns;
for(String pattern:patterns){
     hscan(...);
}
```
In my current example, the number of patterns in 273. And the hscan() query has to be called that many times.

**Comment by deepakpol** [comment]
@chinmaym7430 @sheinbergon assuming when the query performance was compared with SQL it was also for executing match for all patterns...otherwise it indeed seems misplaced 
I see a few possible options to go next step in terms of performance based on your current approach -
1. Try and see if 273 patterns can be combined to form one or fewer patterns. This would reduce the over the network calls and multiple scans currently happening.
2. Use pooling/concurrency
3. I feel 1s is still a longer time to scan 1M keys. Even `KEYS` scan is supposed to take about 40 ms based on redis docs. It might help to fine-tune the performance further.

However, if this is supposed to be a real time use case for user (waiting for the results on a GUI) then it perhaps might make sense to use a search engine?

**Comment by sheinbergon** [comment]
NoSQL K/V Engines are not the best use-case/data-model-choice when it comes to range/pattern scanning. Even solutions like Cassandra/Couchbase/MongoDB that support secondary indices don't perform as well , sometimes require a very specific arrangement of data and/or require multiple queries run against the datasource.

Also note that combining fuzzy queries ( globing/wild card ) with multiple values ( you said you needed to match several hundreds of patterns ) is not supported on all SQL implmentations , and as @deepakpol  mentioned - would be more relevant to a search engine  ( say Elasticsearch ).

http://stackoverflow.com/questions/3014940/is-there-a-combination-of-like-and-in-in-sql

@chinmaym7430 how did you implement your SQL-RDMBS query ?

**Comment by chinmaym7430** [comment]
Hey everyone, thanks for all the help. I finally changed the implementation logic, I realized that I could somehow reduce the number of complete scans. So after some thinking, my code now relies on just getting the values based on particular key search. Searching with patterns in no longer necessary.
And as far as complete scan is concerned, the performance is much better than RDBMs(as expected).
Anyways, thanks for all your help. Learned a lot :)
@sheinbergon Thanks for all the help.
